### PR TITLE
Implement kstatus for health checks

### LINKED ISSUE #68: Extend kustomize controller health assessment to support HelmReleases
If the kustomize controller health assessment was extended to support HelmReleases the kustomize-controller could be used to deploy helmreleases custom resources and detect execution outcome based on HR status. This would enable users to define kustomization custom resources linked by the dependsOn field and sequence deployment of HelmReleases with other object creations.

### PR DESCRIPTION
**Pull request by phillebaba**
This change replaces the existing health check logic with the [kstatus](https://github.com/kubernetes-sigs/cli-utils/tree/master/pkg/kstatus) library.
The purpose for the change is to expand on the possible resource types that can be used in the health checks. As long as the resource implements the standards defined by kstatus it will be compatible with the health check.

Fixes #68

### CODE REVIEW

File: api/v1alpha1/kustomization_types.go [hunk 1]
```diff
@@ -82,10 +83,9 @@ type KustomizationSpec struct {
 
 // WorkloadReference defines a reference to a Deployment, DaemonSet or StatefulSet.
 type WorkloadReference struct {
-	// Kind is the type of resource being referenced.
-	// +kubebuilder:validation:Enum=Deployment;DaemonSet;StatefulSet
+	// GroupKind is the type of resource being referenced.
 	// +required
-	Kind string `json:"kind"`
+	GroupKind metav1.GroupKind `json:"groupKind"`
```
Comment (phillebaba) [inline api/v1alpha1/kustomization_types.go:88]: Need to update the docs with the api change.

File: api/v1alpha1/kustomization_types.go [hunk 2]
```diff
@@ -82,10 +83,9 @@ type KustomizationSpec struct {
 
 // WorkloadReference defines a reference to a Deployment, DaemonSet or StatefulSet.
 type WorkloadReference struct {
```
Comment (phillebaba) [inline api/v1alpha1/kustomization_types.go:85]: Should it still be called WorkloadReference as there no longer is a limit on the type of resource to use for the health checks.

File: controllers/kustomization_controller.go [hunk 3]
```diff
@@ -615,47 +621,90 @@ func (r *KustomizationReconciler) prune(kustomization kustomizev1.Kustomization,
...
+				}
+			case event.ErrorEvent:
+				return e.Error
+				// Event does not behave like expected and will not occur when all the
+				// resources are in a current state.
+				/*case event.CompletedEvent:
+				if alerts != "" && kustomization.Status.LastAppliedRevision != revision {
+					r.event(kustomization, revision, recorder.EventSeverityInfo, alerts)
+				}
+				return nil*/
```
Comment (phillebaba) [inline controllers/kustomization_controller.go:696]: This issue needs to be fixed before the PR can be merged.

I have created an issue for this in  the cli-utils repo.
https://github.com/kubernetes-sigs/cli-utils/issues/232

File: api/v1alpha1/kustomization_types.go (hunk 1 above)
Comment (stefanprodan) [reply api/v1alpha1/kustomization_types.go:88]: Let's make this the same with HPA and Flagger way of referencing objects:

```
// CrossNamespaceObjectReference contains enough information to let you locate the
// typed referenced object at cluster level
type CrossNamespaceObjectReference struct {
	// API version of the referent
	// +optional
	APIVersion string `json:"apiVersion,omitempty"`

	// Kind of the referent
	// +optional
	Kind string `json:"kind,omitempty"`

	// Name of the referent
	Name string `json:"name"`

	// Namespace of the referent
	// +optional
	Namespace string `json:"namespace,omitempty"`
}
```

File: api/v1alpha1/kustomization_types.go (hunk 2 above)
Comment (stefanprodan) [reply api/v1alpha1/kustomization_types.go:85]: See my comment on `CrossNamespaceObjectReference`

File: go.mod [hunk 4]
```diff
@@ -23,7 +23,9 @@ require (
 	google.golang.org/grpc v1.26.0
 	k8s.io/api v0.18.8
 	k8s.io/apimachinery v0.18.8
+	k8s.io/cli-runtime v0.18.8 // indirect
 	k8s.io/client-go v0.18.8
+	sigs.k8s.io/cli-utils v0.19.2
```
Comment (stefanprodan) [inline go.mod:28]: This should match all the other Kubernetes packages `sigs.k8s.io/cli-utils v0.18.8`

File: controllers/kustomization_controller.go [hunk 5]
```diff
@@ -615,47 +621,103 @@ func (r *KustomizationReconciler) prune(kustomization kustomizev1.Kustomization,
...
+			return fmt.Errorf("Health check timed out for [%v]", strings.Join(notHealthy, ", "))
+		case e := <-eventsChan:
+			switch e.EventType {
+			case event.ResourceUpdateEvent:
+				if e.Resource.Status == status.CurrentStatus {
+					healthySet[e.Resource.Identifier.String()] = true
+					r.Log.WithValues(
+						strings.ToLower(kustomization.Kind),
+						fmt.Sprintf("%s/%s", kustomization.GetNamespace(), kustomization.GetName()),
+					).Info(fmt.Sprintf("Health check passed for %s", e.Resource.Identifier.String()))
```
Comment (phillebaba) [inline controllers/kustomization_controller.go:697]: Identifier string should be something a bit more human readable.

File: api/v1alpha1/reference_types.go [hunk 6]
```diff
@@ -24,7 +24,6 @@ type CrossNamespaceObjectReference struct {
 	APIVersion string `json:"apiVersion,omitempty"`
 
 	// Kind of the referent
-	// +kubebuilder:validation:Enum=GitRepository
```
Comment (phillebaba) [inline api/v1alpha1/reference_types.go:27]: This has removed input verification for SourceRef.

Comment (stefanprodan) [reply api/v1alpha1/reference_types.go:27]: Please keep that struct for the source ref but rename it to `CrossNamespaceSourceReference`.

**Comment by phillebaba** [comment]
The changes are still missing doc updates and tests to cover the health check functionality.

**Comment by stefanprodan** [comment]
@phillebaba can you please test it with a HelmRelease and post here the controller logs? Thanks

**Comment by phillebaba** [comment]
The following logs were copied from the KustomizeController during different events.

When the HelmRelease is ready.
```
2020-09-13T13:58:43.307+0200    INFO    controllers.Kustomization       All dependencies area ready, proceeding with reconciliation     {"controller": "kustomization", "request": "gitops-system/webapp-backend"}
2020-09-13T13:58:43.586+0200    INFO    controllers.Kustomization       Kustomization applied in 117.098796ms   {"kustomization": "gitops-system/webapp-backend", "output": {"deployment.apps/backend":"configured","horizontalpodautoscaler.autoscaling/backend":"unchanged","service/backend":"unchanged"}}
2020-09-13T13:58:43.603+0200    INFO    controllers.Kustomization       Reconciliation finished in 301.192168ms, next run in 10m0s      {"controller": "kustomization", "request": "gitops-system/webapp-backend", "revision": "master/d3865f924736fe5d4c52bb9f8db86e46cfd1fe52"}
```

When the HelmRelease is not ready.
```
2020-09-13T15:04:20.465+0200    INFO    controllers.Kustomization       Kustomization applied in 135.188776ms   {"kustomization": "gitops-system/webapp-backend", "output": {"deployment.apps/backend":"configured","horizontalpodautoscaler.autoscaling/backend":"unchanged","service/backend":"unchanged"}}
2020-09-13T15:05:21.468+0200    DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"Kustomization","namespace":"gitops-system","name":"webapp-backend","uid":"6ed5c2fb-466d-4358-978e-ac7cc93a2789","apiVersion":"kustomize.toolkit.fluxcd.io/v1alpha1","resourceVersion":"6568"}, "reason": "error", "message": "Health check timed out for [HelmRelease 'gitops-system/test']"}
2020-09-13T15:05:21.475+0200    INFO    controllers.Kustomization       Reconciliation finished in 1m1.297442713s, next run in 10m0s    {"controller": "kustomization", "request": "gitops-system/webapp-backend", "revision": "master/d3865f924736fe5d4c52bb9f8db86e46cfd1fe52"}
2020-09-13T15:05:21.475+0200    ERROR   controller      Reconciler error        {"reconcilerGroup": "kustomize.toolkit.fluxcd.io", "reconcilerKind": "Kustomization", "controller": "kustomization", "name": "webapp-backend", "namespace": "gitops-system", "error": "Health check timed out for [HelmRelease 'gitops-system/test']"}
github.com/go-logr/zapr.(*zapLogger).Error
        /home/philip/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /home/philip/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:237
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /home/philip/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
        /home/philip/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
        /home/philip/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
        /home/philip/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /home/philip/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
        /home/philip/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:90
```

When the HelmRelease does not exist.
```
2020-09-13T13:23:55.184+0200    INFO    controllers.Kustomization       Reconciliation finished in 258.536169ms, next run in 1h0m0s     {"controller": "kustomization", "request": "gitops-system/webapp-common", "revision": "master/d3865f924736fe5d4c52bb9f8db86e46cfd1fe52"}
2020-09-13T13:23:55.214+0200    INFO    controllers.Kustomization       Kustomization applied in 108.413783ms   {"kustomization": "gitops-system/webapp-backend", "output": {"deployment.apps/backend":"configured","horizontalpodautoscaler.autoscaling/backend":"unchanged","service/backend":"unchanged"}}
2020-09-13T13:24:56.215+0200    DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"Kustomization","namespace":"gitops-system","name":"webapp-backend","uid":"8c387886-b689-427f-932e-020c397ff685","apiVersion":"kustomize.toolkit.fluxcd.io/v1alpha1","resourceVersion":"2244"}, "reason": "error", "message": "Health check timed out for [HelmRelease 'gitops-system/frontend']"}
2020-09-13T13:24:56.223+0200    INFO    controllers.Kustomization       Reconciliation finished in 1m1.297157763s, next run in 10m0s    {"controller": "kustomization", "request": "gitops-system/webapp-backend", "revision": "master/d3865f924736fe5d4c52bb9f8db86e46cfd1fe52"}
2020-09-13T13:24:56.223+0200    ERROR   controller      Reconciler error        {"reconcilerGroup": "kustomize.toolkit.fluxcd.io", "reconcilerKind": "Kustomization", "controller": "kustomization", "name": "webapp-backend", "namespace": "gitops-system", "error": "Health check timed out for [HelmRelease 'gitops-system/frontend']"}
github.com/go-logr/zapr.(*zapLogger).Error
        /home/philip/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /home/philip/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:237
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /home/philip/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
        /home/philip/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
        /home/philip/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
        /home/philip/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /home/philip/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
        /home/philip/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:90
```

I am a bit confused though about how the controller does health checks. I am using the podinfo frontend chart to do my testing. After setting the faults values so that it would never reach a ready state, I was expecting the HelmRelease to not be ready. That does not seem to be the case. Instead the HelmRelease ends up in a healthy state.

Even weirder is that if I set the image to use a tag that does not exist the chart will still be marked as ready. But if I change to another chart like redis from the stable repository and set an incorrect image tag, the helm release will fail as expected.

The problem is probably part of the helm controller so we should probably track it there, and consider this feature to acheive what it set out to do.

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @phillebaba 🎖️

**Pull request merged**